include/pybind11/embedded.h

```diff
@@ -0,0 +1,151 @@
+/*
+    pybind11/embedded.h: Support for embedding the interpreter
+
+    Copyright (c) 2017 Wenzel Jakob <wenzel.jakob@epfl.ch>
+
+    All rights reserved. Use of this source code is governed by a
+    BSD-style license that can be found in the LICENSE file.
+*/
+
+#pragma once
+
+#include "pybind11.h"
+#include "eval.h"
+
+#if defined(PYPY_VERSION)
+#  error Embedding the interpreter is not supported on PyPy
+#endif
+
+#if PY_MAJOR_VERSION >= 3
+#define PYBIND11_ADD_EMBEDDED_MODULE_IMPL(name)        \
+    extern "C" PyObject *pybind11_init_impl_##name() { \
+        auto m = pybind11::module(#name);              \
+        pybind11_init_##name(m);                       \
+        return m.ptr();                                \
+    }
+#else
+#define PYBIND11_ADD_EMBEDDED_MODULE_IMPL(name)   \
+    extern "C" void pybind11_init_impl_##name() { \
+        auto m = pybind11::module(#name);         \
+        pybind11_init_##name(m);                  \
+    }
+#endif
+
+/** \rst
+    Add a new module to the table of builtins for the interpreter. Must be
+    defined in global scope.
+
+    .. code-block:: cpp
+
+        PYBIND11_ADD_EMBEDDED_MODULE(example)(py::module &m) {
+            // ... initialize functions and classes here
+            m.def("foo", []() {
+                return "Hello, World!";
+            });
+        }
+ \endrst */
+#define PYBIND11_ADD_EMBEDDED_MODULE(name)                                        \
```

Comment:
Should it be called ``PYBIND11_EMBEDDED_PLUGIN``? (the module<->plugin naming choice is a bit unfortunate in retrospect, but it might be better to stay consistent)